doc/command_guidelines.md

```diff
@@ -0,0 +1,63 @@
+# Command Guidelines
+
+This document describes the command guidelines for 'az' and applies to both CLI command modules and extensions.
+
+Guidelines marked (*) only apply to command modules, not extensions.
+
+If in doubt, ask!
+
+## General Patterns
+
+- Be consistent with POSIX tools (support piping, work with grep, awk, jq, etc.)
+- Support tab completion for parameter names and values (e.g. resource names)
+- Commands must follow a "[noun] [noun] [verb]" pattern
+- For nouns that only support a single verb, the command should be named as a single hyphenated verb-noun pair
+- All commands, command group and arguments must have descriptions
+- You must provide command examples for non-trivial commands
+- All commands must support all output types (json, tsv, table)
+- Provide custom table outputs for commands that don't provide table output automatically
+- Commands must return an object or dictionary (do not string, Boolean, etc. types)
+- Use `stdout` and `stderr` appropriately.
+- Command output must go to stdout, everything else to stderr (log/status/errors).
+- Log to `logger.error()` or `logger.warning()` for user messages; do not use the `print()` function
+- Use the appropriate logging level for printing strings. e.g. `logging.info(“Upload of myfile.txt successful”)` NOT `return “Upload successful”`.
+
+## Command Naming and Behavior Guidance
+
+- Multi-word subgroups should be hyphenated
+e.g. `foo-resource` instead of `fooresource`
+- All command names should contain a verb
+e.g. `account get-connection-string` instead of `account connection-string`
+- Avoid hyphenated command names when moving the commands into a subgroup would eliminate the need.
+e.g. `database show` and `database get` instead of `show-database` and `get-database`
+- If a command subgroup would only have a single command, move it into the parent command group and hyphenate the name. This is common for commands which exist only to pull down cataloging information.
+e.g. `database list-sku-definitions` instead of `database sku-definitions list`
+- Avoid command subgroups that have no commands. This often happens at the first level of a command branch.
+e.g. `keyvault create` instead of `keyvault vault create` (where `keyvault` only has subgroups and adds unnecessary depth to the tree).
```

Comment:
I prefer the original, even though it is longer. The example would likely not be helpful at all to someone who isn't us without the preserved help output. When I read this, it sounds like it's keyvault specific and misses the bigger picture.

Reply:
added to an expandable section.